R/featurization/get_features.R

```diff
@@ -0,0 +1,250 @@
+source('glove_analysis.R')
+source('bias_themes_analysis.R')
+
+registerDoParallel(detectCores())
+
+#
+# Get the raw text data for embedding training
+# 
+
+b_corp <- fread('C:/users/ben_ryan/documents/darpa ugb/breitbart_texts.csv',
+                header=FALSE)
+
+bv = trainEmbeddings(b_corp$V2, alpha=0.05)
+write.table(bv,
+            'breitbart_embeddings_wComments.txt',
+            col.names=FALSE,
+            quote=FALSE,
+            fileEncoding='utf-8')
+
+# To read in again from saved:
+# bv = fread('breitbart_embeddings_wComments.txt', header=FALSE)
+# terms <- bv$V1
+# bv[,V1:=NULL]
+# bv <- data.matrix(bv)
+# rownames(bv) <- terms
+
+#
+#
+# Begin feature extraction from articles and comments datasets
+#
+# 
+
+barts = fread('breitbart_articles.csv') #load breitbart articles
+bcs = fread('breitbart_comments.csv') #load breitbart comments
+
+# 
+# Derive bias component vectors
+# 
+
+gender_bias <- deriveBias(genderPairs, wv=bv, method='pca', diag=TRUE)
+race_bias <- deriveBias(adj_race_names, wv=bv, diag=TRUE)
+power_bias <- deriveBias(powerPairs, wv=bv, method='pca', diag=TRUE)
+
+#
+# Derive embeddings for articles and article titles, and merge into one 
+# data.table. 
+# 
+
+arts_text_tokens = getTokens(barts$art_text)
+bgloved_arts <- lapply(arts_text_tokens, gloved, wv=bv)
+bgloved_arts <- data.table(do.call(rbind, bgloved_arts))
+
+# 
+# Propagate bias components
+#
+
+gender <- foreach(i = 1:nrow(bgloved_arts),
+                  .inorder=TRUE,
+                  .export=c('biasComponent','avgBias','bv','gender_bias','power_bias','race_bias'),
+                  .packages = c('NLP','openNLP','dplyr'),
+                  .multicombine = TRUE, 
+                  .options.multicore = list(preschedule = FALSE)) %dopar% {
+            avgBias(arts_text_tokens[[i]], gender_bias, bv)
+        }
+
+bgloved_arts$gender <- unlist(gender)
+
+
+race <- foreach(i = 1:nrow(bgloved_arts),
+                  .inorder=TRUE,
+                  .multicombine = TRUE, 
+                  .options.multicore = list(preschedule = FALSE)) %dopar% {
+            avgBias(arts_text_tokens[[i]], race_bias, bv)
+        }
+
+bgloved_arts$race <- unlist(race)
+
+
+power <- foreach(i = 1:nrow(bgloved_arts),
+                  .inorder=TRUE,
+                  .multicombine = TRUE, 
+                  .options.multicore = list(preschedule = FALSE)) %dopar% {
+            avgBias(arts_text_tokens[[i]], power_bias, bv)
+        }
+
+bgloved_arts$power <- unlist(power)
+
+# 
+# Set column names and add article id
+#
+
+names(bgloved_arts) <- paste('art_text_embeddings_',names(bgloved_arts), sep='')
+bgloved_arts$art_id <- barts$art_id
+
+#
+# Add in the article metadata (sentiment scores and number of comments per article)
+# 
+
+sent_cols <- c(names(barts)[names(barts) %like% 'sent_'], 'art_comments', 'art_id')
+bgloved_arts <- merge(bgloved_arts, barts[,sent_cols,with=FALSE], by='art_id')
+
+#
+# Okay, now article titles, same deal
```

Comment:
It _seems_ the code is basically the same, other than the input vector, so could this be made into a function that takes the name of the input vector to reduce the amount of repetitious code? That said, if I missed something unique about the difference between article text, article title, and comments, then leaving it more explicit like this can be fine.